Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h

```diff
@@ -95,24 +96,32 @@ class VideoMediaSampleRenderer : public ThreadSafeRefCountedAndCanMakeThreadSafe
 
     void resetReadyForMoreSample();
     void initializeDecompressionSession();
+    void enqueueSample(RetainPtr<CMSampleBufferRef>&&, int);
     void decodeNextSample();
     void decodedFrameAvailable(RetainPtr<CMSampleBufferRef>&&);
+    void maybeQueueFrameForDisplay(CMSampleBufferRef);
     void flushCompressedSampleQueue();
     void flushDecodedSampleQueue();
     void purgeDecodedSampleQueue();
     CMBufferQueueRef ensureDecodedSampleQueue();
     void assignResourceOwner(CMSampleBufferRef);
     void maybeBecomeReadyForMoreMediaData();
 
+    void cancelTimer();
+
     const Ref<WTF::WorkQueue> m_workQueue;
     RetainPtr<AVSampleBufferDisplayLayer> m_displayLayer;
     RetainPtr<AVSampleBufferVideoRenderer> m_renderer;
-    RetainPtr<CMTimebaseRef> m_timebase;
-    RetainPtr<CMBufferQueueRef> m_compressedSampleQueue; // created on the main thread, always set if m_decompressionSession and on workQueue.
-    RetainPtr<CMBufferQueueRef> m_decodedSampleQueue;
-    OSObjectPtr<dispatch_source_t> m_timerSource;
+    mutable Lock m_lock;
+    RetainPtr<CMTimebaseRef> m_timebase WTF_GUARDED_BY_LOCK(m_lock);
+    OSObjectPtr<dispatch_source_t> m_timerSource WTF_GUARDED_BY_LOCK(m_lock);
+    std::atomic<ssize_t> m_framesBeingDecoded { 0 };
+    std::atomic<int> m_flushId { 0 };
+    Deque<std::pair<RetainPtr<CMSampleBufferRef>, int>> m_compressedSampleQueue WTF_GUARDED_BY_CAPABILITY(m_workQueue.get());
+    RetainPtr<CMBufferQueueRef> m_decodedSampleQueue WTF_GUARDED_BY_CAPABILITY(m_workQueue.get());
     RefPtr<WebCoreDecompressionSession> m_decompressionSession;
     bool m_isDecodingSample { false };
+    std::atomic<bool> m_isDisplayingSample { false };
```

Comment:
Isn't this one always checked/set from the work queue like m_decodedSampleQueue?

Reply:
yes indeed. it's a remnant of an earlier code where it was checked before the call to copyPixelBuffer to avoid the sync dispatch. I'll remove it. That code is re-introduced in another PR